### PR TITLE
Fix crash when invalid args for library polished's functions

### LINKED ISSUE #6855: Gradients in theme break shortcut menu
**Describe the bug**
Whenever one uses a gradient in the theme, the shortcut button will cause the entire app to crash. 
**To Reproduce**
Theme with crash: 
```js
import { create } from '@storybook/theming';

const blue = '#0063ba';
const grey = '#e6e6e6';
const white = '#ffffff';

export default create({
  base: 'light',

  colorPrimary: blue,
  colorSecondary: blue,

  // UI
  appBg: `linear-gradient(to bottom, ${grey}, ${white})`,
  appContentBg: white,
  appBorderColor: blue,
  appBorderRadius: 0,
});
```
---
Theme without crash:
```js
import { create } from '@storybook/theming';

const blue = '#0063ba';
const grey = '#e6e6e6';
const white = '#ffffff';

export default create({
  base: 'light',

  colorPrimary: blue,
  colorSecondary: blue,

  // UI
  appBg: white,
  appContentBg: white,
  appBorderColor: blue,
  appBorderRadius: 0,
});
```
**Expected behavior**
The gradient background works, but I also expect the shortcut menu to open normally.

**Screenshots**
A screenshot would simply be a white page.

**System:**
 - OS: [MacOS]
 - Device: [Macbook Pro 2018]
 - Browser: [chrome]
 - Framework: [react]
 - Version: [5.0.6]

**Additional**
Also discussed here: #6756 
I think a better error is required to prevent confusion with the theme.

### PR DESCRIPTION
**Pull request by lonyele**
Issue: Fixes #6855, #6756, #6135 

#### Problem 
When using some functions from [polished](https://github.com/styled-components/polished) such as `lighten(...), darken(...)`, its arguments that can't be converted to RGB [throws the error](https://github.com/styled-components/polished/blob/master/src/color/parseToRgb.js#L123) 

Commit that possibly introduced this [bug](https://github.com/storybookjs/storybook/pull/5562/commits/9a50e803e39006ece5718fc7b896143f07e6b63e#diff-0d0c72f854c189e6d0d17a1948570d5dR39)

#### Implementation
- Check if arguments of those functions are valid(gradient, css variable, calc etc...)
- Since it is used on different places, I put it into `util.js` from `@storybook/theming` (I couldn't find the place to put some util/shared functionalities across packages. If it doesn't look ok, please let me know where to put these kinds of functions)

## How to test

- Is this testable with Jest or Chromatic screenshots?
- Does this need a new example in the kitchen sink apps? No
- Does this need an update to the documentation? No

### CODE REVIEW

**Review by shilman — DISMISSED** [review]
Nice!!! Can you call these `lighten` and `darken` for consistency's sake?

**Comment by lonyele** [comment]
Yeap, I applied your feedbacks~

File: lib/theming/src/utils.ts [hunk 1]
```diff
@@ -1 +1,26 @@
+import { rgba, lighten, darken } from 'polished';
+
 export const mkColor = (color: string) => ({ color });
+
+// Passing arguments that can't be converted to RGB such as linear-gradient
+// to library polished's functions such as lighten or darken throws the error
+// that crashes the entire storybook. It needs to be guarded when arguments
+// of those functions are from user input.
+const isLinearGradient = (color: string) => {
+  return typeof color === 'string' && color.includes('linear-gradient');
```
Comment (ndelangen) [inline lib/theming/src/utils.ts]: I think this function is too specific, linear-gradients aren't the only type of gradients.

How about:

```js
const isColorVarChangeable = (color: string) => {
  return !!color.match(/(gradient|var)/);
}
```

related to #6135

**Review by ndelangen — CHANGES_REQUESTED** [review]

**Comment by ndelangen** [comment]
Thanks for contributing @lonyele 🎉

File: lib/theming/src/utils.ts (hunk 1 above)
Comment (lonyele) [reply lib/theming/src/utils.ts]: Oh, you are right... I only thought about the linear-gradient even though I wrote a comment `things that can't be converted to RGB`  I change to your function! Please check

Comment (ndelangen) [reply lib/theming/src/utils.ts]: Could you add a few test cases for the function?

**Comment by lonyele** [comment]
Oh, I'm working on this one, please wait for a while

**Comment by lonyele** [comment]
Hm... I'm not sure if the snapshot should be updated(https://circleci.com/gh/storybookjs/storybook/130895?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link) Help is needed here...

I found from library `polished` that if anything goes wrong, it throws [`new PolishedError(...)`](https://github.com/styled-components/polished/search?q=Error&type=Code)  I feel like this is a small hot fix, but it is a not bad start to be safe from `polished`'s exception. Potentially any user inputs that are being passed to `polished` can go wrong. I'm not sure if it should be refactored since there are a lot and it is only for the few cases, but if it is needed then any custom logic can be added easily [here](https://github.com/storybookjs/storybook/pull/6939/files#diff-83aa48423f4a4a8b26b8eb4786688a9fR29)

File: lib/theming/src/utils.ts (hunk 1 above)
Comment (lonyele) [reply lib/theming/src/utils.ts]: @ndelangen I wrote some tests and it worked fine, but I'm not sure about updating [the snapshot test](https://circleci.com/gh/storybookjs/storybook/130895?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link)

**Review by lonyele — COMMENTED** [review]
[sorry mistake](https://github.com/storybookjs/storybook/pull/6939/files#diff-83aa48423f4a4a8b26b8eb4786688a9fR54)

File: lib/theming/src/utils.ts [hunk 2]
```diff
@@ -1 +1,59 @@
...
+
+  if (!isValidColorForPolished(color)) {
+    return color;
+  }
+
+  // Guard anything that is not working with polished.
+  try {
+    return applyPolished(type, color);
+  } catch (error) {
+    return color;
```
Comment (lonyele) [inline lib/theming/src/utils.ts]: Logging a message can maybe helpful to the users or tracking bug later, but I'm not sure

Comment (ndelangen) [reply lib/theming/src/utils.ts]: Yeah, we'll see if an issue opens up, let's keep it not logging anything for now.

**Comment by ndelangen** [comment]
@lonyele run:
```sh
yarn test --core --update
```

Then commit the changes generated, to fix the snapshots.

**Comment by lonyele** [comment]
@ndelangen Hm... my changes of chromatic looks very wrong. I'll investigate this

**Comment by ndelangen** [comment]
Indeed, I denied the changes that seems wrong.

**Comment by lonyele** [comment]
@ndelangen  oh god... one of the most stupid mistake I've made in few months. I hung my head in shame...
Just a side note, because of [this try/catch](https://github.com/storybookjs/storybook/pull/6939/files#diff-83aa48423f4a4a8b26b8eb4786688a9fR51) tests were all passed even with this stupid mistake.

**Review by ndelangen — APPROVED** [review]

**Pull request merged**